cmd/core/service/ipam.go

```diff
@@ -0,0 +1,127 @@
+package service
+
+import (
+	"errors"
+	"net"
+	"strconv"
+	"sync"
+)
+
+type IPAM struct {
+	freeIPs   []net.IP
+	busyIPs   map[uint64]net.IP
+	freeTEIDs []uint32
+	busyTEIDs map[string]uint32
+	sync.RWMutex
+}
+
+func NewIPAM(ipRange string) (*IPAM, error) {
+	_, ipNet, err := net.ParseCIDR(ipRange)
+	if err != nil {
+		return nil, err
+	}
+
+	freeIPs := make([]net.IP, 0, 255)
+	busyIPs := make(map[uint64]net.IP)
+
+	freeTEIDs := make([]uint32, 0, 255)
+	busyTEIDs := make(map[string]uint32)
+
+	var teid uint32
+	ip := ipNet.IP
+
+	for ipNet.Contains(ip) {
+		freeIPs = append(freeIPs, net.IP(ip))
+		ip = nextIP(ip)
+		teid++
+		freeTEIDs = append(freeTEIDs, teid)
+
+	}
+
+	return &IPAM{
+		freeIPs:   freeIPs,
+		busyIPs:   busyIPs,
+		freeTEIDs: freeTEIDs,
+		busyTEIDs: busyTEIDs,
+	}, nil
+}
+
+func (ipam *IPAM) AllocateIP(key uint64) (net.IP, error) {
+	ipam.Lock()
+	defer ipam.Unlock()
+
+	if len(ipam.freeIPs) > 0 {
+		ip := ipam.freeIPs[0]
+		ipam.freeIPs = ipam.freeIPs[1:]
+		ipam.busyIPs[key] = ip
+		return ip, nil
+	} else {
+		return nil, errors.New("no free ip available")
+	}
+}
+
+func (ipam *IPAM) AllocateTEID(seID uint64, pdrID uint16, chooseID uint8) (uint32, error) {
+	ipam.Lock()
+	defer ipam.Unlock()
+
+	if len(ipam.freeTEIDs) > 0 {
+		key := strconv.Itoa(int(seID)) + ":" + strconv.Itoa(int(pdrID))
+		if chooseID != 0 {
+			key += ":" + strconv.Itoa(int(chooseID))
```

Comment:
`chooseID` can't be a part of key session-wide because it is unique only within a corresponding request.